### PR TITLE
#minor Node executions

### PR DESCRIPTION
**Pull request by pmahindrakar-oss**
# TL;DR

Adds node execution details on the execution view

Gets the details in the tree view. This is the default view for details which fetch the node exeuction details and task executions on each node. 

1] Default is tree view and table format is not supported for this explicitly
```
flytectl get execution -p flytectldemo -d development f3a5a4034960f4aa1a09 --details

TABLE format is not supported on detailed view and defaults to tree view. Choose either json/yaml
Node Executions
└── start-node - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
└── n1 - SUCCEEDED - 2021-06-30 08:51:07.3111846 +0000 UTC - 2021-06-30 08:51:17.192852 +0000 UTC
│   ├── Attempt :0
│       └── Task - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
│       └── Task Type - 
│       └── Reason - 
│       └── Logs :
│           └── Name :Kubernetes Logs (User)
│           └── URI :http://localhost:30082/#/log/flytectldemo-development/f3a5a4034960f4aa1a09-n1-0/pod?namespace=flytectldemo-development
└── n0 - SUCCEEDED - 2021-06-30 08:49:01.582546 +0000 UTC - 2021-06-30 08:51:07.1850111 +0000 UTC
│   ├── Attempt :0
│       └── Task - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
│       └── Task Type - 
│       └── Reason - 
│       └── Logs :
│           └── Name :Kubernetes Logs (User)
│           └── URI :http://localhost:30082/#/log/flytectldemo-development/f3a5a4034960f4aa1a09-n0-0/pod?namespace=flytectldemo-development
└── end-node - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
```

2] Yaml  view for the details

```
flytectl get execution -p flytectldemo -d development f3a5a4034960f4aa1a09 --details  -o yaml
- closure:
    createdAt: "2021-06-30T08:51:17.271218200Z"
    phase: SUCCEEDED
    updatedAt: "2021-06-30T08:51:17.311695100Z"
  id:
    executionId:
      domain: development
      name: f3a5a4034960f4aa1a09
      project: flytectldemo
    nodeId: end-node
  inputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/end-node/data/inputs.pb
  metadata:
    specNodeId: end-node
- closure:
    createdAt: "2021-06-30T08:49:01.528398300Z"
    duration: 125.602465100s
    outputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/n0/data/0/outputs.pb
    phase: SUCCEEDED
    startedAt: "2021-06-30T08:49:01.582546Z"
    updatedAt: "2021-06-30T08:51:07.185011100Z"
  id:
    executionId:
      domain: development
      name: f3a5a4034960f4aa1a09
      project: flytectldemo
    nodeId: n0
  inputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/n0/data/inputs.pb
  metadata:
    specNodeId: n0
- closure:
    createdAt: "2021-06-30T08:51:07.249937800Z"
    duration: 9.881667400s
    outputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/n1/data/0/outputs.pb
    phase: SUCCEEDED
    startedAt: "2021-06-30T08:51:07.311184600Z"
    updatedAt: "2021-06-30T08:51:17.192852400Z"
  id:
    executionId:
      domain: development
      name: f3a5a4034960f4aa1a09
      project: flytectldemo
    nodeId: n1
  inputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/n1/data/inputs.pb
  metadata:
    specNodeId: n1
- closure:
    createdAt: "2021-06-30T08:49:01.491365700Z"
    phase: SUCCEEDED
    updatedAt: "2021-06-30T08:49:01.491365700Z"
  id:
    executionId:
      domain: development
      name: f3a5a4034960f4aa1a09
      project: flytectldemo
    nodeId: start-node
  inputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/start-node/data/inputs.pb
  metadata:
    specNodeId: start-node
```
3] Default tree view for task executions in a node provide a name of the node.

```
flytectl get execution -p flytectldemo -d development f3a5a4034960f4aa1a09 --details  --nodeId n0
TABLE format is not supported on detailed view and defaults to tree view. Choose either json/yaml
Task Executions
└── Attempt :0
    └── Task - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
    └── Task Type - 
    └── Reason - 
    └── Logs :
        └── Name :Kubernetes Logs (User)
        └── URI :http://localhost:30082/#/log/flytectldemo-development/f3a5a4034960f4aa1a09-n0-0/pod?namespace=flytectldemo-development
```
4] Yaml representation for the same task executions.

```
flytectl get execution -p flytectldemo -d development f3a5a4034960f4aa1a09 --details  --nodeId n0 -o yaml
closure:
  createdAt: "2021-06-30T08:49:01.573698800Z"
  logs:
  - messageFormat: JSON
    name: Kubernetes Logs (User)
    uri: http://localhost:30082/#/log/flytectldemo-development/f3a5a4034960f4aa1a09-n0-0/pod?namespace=flytectldemo-development
  outputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/n0/data/0/outputs.pb
  phase: SUCCEEDED
  updatedAt: "2021-06-30T08:51:07.169958100Z"
id:
  nodeExecutionId:
    executionId:
      domain: development
      name: f3a5a4034960f4aa1a09
      project: flytectldemo
    nodeId: n0
  taskId:
    domain: development
    name: core.control_flow.subworkflows.t1
    project: flytectldemo
    resourceType: TASK
    version: v3
inputUri: s3://my-s3-bucket/metadata/propeller/flytectldemo-development-f3a5a4034960f4aa1a09/n0/data/inputs.pb
```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1124

## Follow-up issue
_NA_
OR
_https://github.com/lyft/flyte/issues/<number>_

### CODE REVIEW

**Comment by kumare3** [comment]
I am ok with us moving forward with this, we will need a `flytectl get executions execID --nodeId "x"` command I think

**Comment by katrogan** [comment]
hey @pmahindrakar-oss do you have an example of a successful node execution view? 

also how does this work with subworkflows? will you expand the dynamic workflow structure in this visualization too?

just to clarify, Task type and reason are not being returned in Get/ListTaskExecution responses? and what is wrong with the timestamps, are they empty too? can we file an issue to track?

**Comment by pmahindrakar-oss** [comment]
Ok the UI populates the task type using the taskTemplate which is not available in execution, so we might have to do a get task to do this ? Is that right . We might want to fix this in admin according to me.
```
function createTaskNodeExecutionDetails(
    taskTemplate: TaskTemplate
): NodeExecutionDetails {
    return {
        taskTemplate,
        displayId: taskTemplate.id.name,
        displayType:
            taskTypeToNodeExecutionDisplayType[taskTemplate.type as TaskType] ??
            NodeExecutionDisplayType.UnknownTask
    };
}
```
@katrogan  which api can return the taskType and reason.


This is a successful one  . The timestamps are coming as empty over there too , So i have now used createdAt and updatedAt values  at the task level but they don't seems to have any values in start-node and end-node(At the node level the startAt and duration is available)

```
Node Executions
└── end-node - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
└── n0 - SUCCEEDED - 2021-04-14 19:30:05.012270315 +0000 UTC - 2021-04-14 19:30:38.307529333 +0000 UTC
│   ├── Attempt :0
│       └── Task - SUCCEEDED - 2021-04-14 19:30:04.998087892 +0000 UTC - 2021-04-14 19:30:38.226590983 +0000 UTC
│       └── Task Type - 
│       └── Reason - 
│       └── Logs :
│           └── Name :Cloudwatch Logs (User)
│           └── URI :https://console.aws.amazon.com/cloudwatch/home?region=us-east-2#logEventViewer:group=/aws/containerinsights/flyte-demo-2/application;stream=var.log.containers.maz1pw2taq-n0-0_flytetester-development_maz1pw2taq-n0-0-ba2bdce5736ca38ff7c649c36b659f7aafa2e9a50144708d570c74dd2944a408.log
└── start-node - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
```

For subworkflows i am trying to run on nuclyde and getting stuck running those .
https://demo.nuclyde.io/console/projects/flytectldemo/domains/development/workflows/core.control_flow.subworkflows.my_subwf
I have also increaed the memory and cpu . Is there anything else we need to do to schedule executions for these.
```
defaults:
  cpu: "1"
  memory: "150Mi"
limits:
  cpu: "2"
  memory: "450Mi"
```
Seem to be simple enough workflow to require more than this memory or cpu. For subworkflows i have to figure out the data . 
Does it relate to this metadata structure for the node.
```
type NodeExecutionMetaData struct {
	......
        // Boolean flag indicating if the node has child nodes under it
	IsParentNode bool `protobuf:"varint,2,opt,name=is_parent_node,json=isParentNode,proto3" json:"is_parent_node,omitempty"`
       .........
```

yes i am planning to add this` flytectl get executions execID --nodeId "x"` but for 
`-o yaml should return a full Execution Closure which includes node executions` , the current printer only accepts one protobuf message type slice,  Can we create a dynamic proto structure to include node executions being part of the executions ?

**Comment by kumare3** [comment]
@pmahindrakar-oss UI joins with TaskTemplate, but you dont need to - you can simply use Info in the closure - https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/task_execution.proto#L105-L108

**Comment by kumare3** [comment]
@katrogan is this a question `just to clarify, Task type and reason are not being returned in Get/ListTaskExecution responses?` or an explanation?

**Comment by pmahindrakar-oss** [comment]
@katrogan  heres how it looks for subworkflow

```
Node Executions
└── end-node - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
└── n0 - SUCCEEDED - 2021-06-25 06:25:12.547887811 +0000 UTC - 2021-06-25 06:25:21.130349059 +0000 UTC
│   ├── Attempt :0
│       └── Task - SUCCEEDED - 2021-06-25 06:25:12.542144177 +0000 UTC - 2021-06-25 06:25:21.078492416 +0000 UTC
│       └── Task Type - 
│       └── Reason - 
│       └── Logs :
│           └── Name :Cloudwatch Logs (User)
│           └── URI :https://console.aws.amazon.com/cloudwatch/home?region=us-east-2#logEventViewer:group=/aws/containerinsights/flyte-demo-2/application;stream=var.log.containers.f13e09f039c2b4ee281e-n0-0_flyteexamples-development_f13e09f039c2b4ee281e-n0-0-212b2880a7538d00dfa7479c500cbcfcf5ef379a1fb1a313f358bdc8261fbf64.log
└── n1 - SUCCEEDED - 2021-06-25 06:25:21.365050403 +0000 UTC - 2021-06-25 06:25:31.114826925 +0000 UTC
│   ├── Attempt :0
│       └── Task - SUCCEEDED - 2021-06-25 06:25:21.359684419 +0000 UTC - 2021-06-25 06:25:31.078614004 +0000 UTC
│       └── Task Type - 
│       └── Reason - 
│       └── Logs :
│           └── Name :Cloudwatch Logs (User)
│           └── URI :https://console.aws.amazon.com/cloudwatch/home?region=us-east-2#logEventViewer:group=/aws/containerinsights/flyte-demo-2/application;stream=var.log.containers.f13e09f039c2b4ee281e-n1-0_flyteexamples-development_f13e09f039c2b4ee281e-n1-0-28c3d7d9470bb4655bee1961e1eab49b1e9638d700d453a98d7fd50f180abae8.log
└── start-node - SUCCEEDED - 1970-01-01 00:00:00 +0000 UTC - 1970-01-01 00:00:00 +0000 UTC
```

**Comment by katrogan** [comment]
@pmahindrakar-oss is this still a wip? also what are you thoughts about showing node executions in chronological order, rather than reverse as in your example?

**Comment by kumare3** [comment]
> @pmahindrakar-oss is this still a wip? also what are you thoughts about showing node executions in chronological order, rather than reverse as in your example?

haha, thats a good catch @katrogan - @pmahindrakar-oss it should be in the order of created at I think :)

**Comment by pmahindrakar-oss** [comment]
Thanks @katrogan and @katrogan . Yeah i will update these details and ask you guyz for re-review. WIP for unit tests

File: cmd/get/execution.go [hunk 1]
```diff
@@ -49,6 +51,31 @@ Retrieves all the execution within project and domain in json format.
 
  bin/flytectl get execution -p flytesnacks -d development -o json
 
+
+Get more details for the execution using --details flag which shows node executions along with task executions on them.Default view is tree view and TABLE format is not supported on this view
```
Comment (katrogan) [inline cmd/get/execution.go:55]:
```suggestion
Get more details for the execution using --details flag which shows node executions along with task executions on them. Default view is tree view and TABLE format is not supported on this view
```

File: pkg/util/version.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
...om/flyte-bot","followers_url":"https://api.github.com/users/flyte-bot/followers","following_url":"https://api.github.com/users/flyte-bot/following{/other_user}","gists_url":"https://api.github.com/users/flyte-bot/gists{/gist_id}","starred_url":"https://api.github.com/users/flyte-bot/starred{/owner}{/repo}","subscriptions_url":"https://api.github.com/users/flyte-bot/subscriptions","organizations_url":"https://api.github.com/users/flyte-bot/orgs","repos_url":"https://api.github.com/users/flyte-bot/repos","events_url":"https://api.github.com/users/flyte-bot/events{/privacy}","received_events_url":"https://api.github.com/users/flyte-bot/received_events","type":"User","site_admin":false},"content_type":"application/zip","state":"uploaded","size":9390673,"download_count":0,"created_at":"2021-06-30T19:17:03Z","updated_at":"2021-06-30T19:17:04Z","browser_download_url":"https://github.com/flyteorg/flytectl/releases/download/v0.1.29/flytectl_Windows_x86_64.zip"}],"tarball_url":"https://api.github.com/repos/flyteorg/flytectl/tarball/v0.1.29","zipball_url":"https://api.github.com/repos/flyteorg/flytectl/zipball/v0.1.29","body":"## Changelog\n\nd6b262a Added exec command in sandbox (#122)\n\n"}
```
Comment (katrogan) [inline pkg/util/version.yaml:1]: what's the purpose of this file?

**Review by katrogan — COMMENTED** [review]
lgtm, will let @kumare3 review too

File: cmd/get/execution.go (hunk 1 above)
Comment (pmahindrakar-oss) [reply cmd/get/execution.go]: Done

File: pkg/util/version.yaml (hunk 2 above)
Comment (pmahindrakar-oss) [reply pkg/util/version.yaml]: Was an accidental checkin. Thanks for pointing it out. Removed it

**Review by kumare3 — APPROVED** [review]
